Scopes

samandmoore opened the pull request:
cc @smudge

not sure if i want this feature, but if i do, i think i might prefer this implementation over defining the scope method on an instance of `Relation`. thoughts?

samandmoore committed:
baseline support for scopes, if we want that kinda thing

samandmoore committed:
some proof that scopes do what they're supposed to

smudge commented:
Just define all the scopes on `Object` 😉

samandmoore committed:
more tests for the other stuff that i care about

smudge commented on lib/drowsy/relation.rb:
```diff
@@ -73,6 +73,18 @@ def to_http_request
+    end
+  end
+
+  def respond_to?(name, include_private = false)
```
I think this should be [respond_to_missing?](http://blog.marc-andre.ca/2010/11/15/methodmissing-politely/)

smudge commented on lib/drowsy/relation.rb:
```diff
@@ -73,6 +73,18 @@ def to_http_request
     Drowsy::HttpRequest.new(connection, http_method, uri_template, params)
   end
 
+  def method_missing(name, *args, &block)
```
Some thoughts, since you tagged me.

Reasons I tend to avoid method missing include:

- It doesn't give you as much visibility into the API via things like `public_methods`.
- Benchmarks are inherently slower than with `define_method`, etc.
- Sometimes encourages catch-all objects whose method_missing implementations handle multiple use cases at once.

Where I tend to use method_missing is in cases where defining the methods up-front would make the API surface massive, or where it really wouldn't be possible to know what the methods will be in advance (e.g. Her's use case with dynamic JSON attributes from an API).

As-is this will totally accomplish what you need, but keep in mind that since scopes are in fact finite and known in advance, there would be many other ways to actually define them as true methods on something.

samandmoore replied:
Yea. I have the same feelings about method missing, I just wanted to see if I could do this without defining the scope method on an instance of Relation which is another type of thing I generally try to avoid. I may reimplement now that I have the tests in there.

smudge replied:
Makes sense! 👍

smudge reviewed (commented):
Since you tagged me, and I had some time during a cab ride, here are the things that came to mind.

samandmoore committed:
respond_to_missing instead

samandmoore committed:
use define_singleton_method on a memoized Relation instance

this requires a slight refactor. moved all to Relation and renamed the
all in scoping to blank relation. this technique only actually works
because we're using clone when we produce a new Relation during
chaining. clone copies the singleton class of the instance but dup does
not.

samandmoore commented on lib/drowsy/scoping.rb:
```diff
@@ -6,14 +6,20 @@ module Drowsy::Scoping
+
+    def blank_relation
+      @blank_relation ||= Drowsy::Relation.new(self)
     end
```
@smudge so, i learned that this strategy of defining a singleton method on a blank relation instance only works because this instance is memoized AND the chaining in Relation uses `clone` to produce new instances. `clone` maintains the singleton class.

smudge replied:
Yeah -- `clone` is key.

samandmoore committed:
Merge remote-tracking branch 'origin/master' into scopes

* origin/master:
  allow Model.{get/post/put/patch/delete}
